### PR TITLE
Optimize Fetching of Eth1 Blocks

### LINKED ISSUE #7375: Error connecting to infura right before proposing a block
This happened at least twice for me and another Discord user reports the same twice. 

Getting this error in the beacon-chain
```
[2020-09-29 11:26:51] ERROR rpc/validator: Failed to get block number from timestamp error=Post "https://goerli.infura.io/v3/a2195fc1addresseditedad5dd401270a355f": context deadline exceeded
[2020-09-29 11:26:51]  WARN rpc/validator: Beacon Node is no longer connected to an ETH1 chain, so ETH1 data votes are now random.
[2020-09-29 11:26:51]  INFO blockchain: Synced new block block=0x3172d6ba... epoch=22 finalizedEpoch=0 finalizedRoot=0x00000000... slot=733 slotInEpoch=29
```
The next slot shown is 733 which is where my validator `0x85b1fd7f48aa3247a540524e0bf30c0c3af9813c7f221f54a949d249a46f41b31ff1dc11f0eaa877c1f13245c8452d16` proposed a block.

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Optimization

**What does this PR do? Why is it needed?**

- [x] BlockByTimestamp is our most expensive method by far when proposing a block. When determining the required eth1
data to vote on, we call on this method. Our previous implementation was very expensive, in some cases it was not bounded 
and would send (eth1_follow_distance) eth1 requests in the event no headers were cached, which is expensive.
- [x] We use an optimized search algorithm by estimating the block by the timestamp difference and then readjusting our parameters to come closer to the required block.
- [x] Add unit tests for new methods added in.

**Which issues(s) does this PR fix?**

Fixes #7375 

**Other notes for review**

### CODE REVIEW

File: beacon-chain/powchain/block_reader.go [hunk 1]
```diff
@@ -84,41 +94,115 @@ func (s *Service) BlockTimeByHeight(ctx context.Context, height *big.Int) (uint6
...
+	}
+	headNumber := big.NewInt(int64(s.latestEth1Data.BlockHeight))
+	headTime := s.latestEth1Data.BlockTime
+	numOfBlocks := uint64(0)
+	estimatedBlk := headNumber.Uint64()
+	maxTimeBuffer := searchThreshold * params.BeaconConfig().SecondsPerETH1Block
+	// Terminate if we cant find an acceptible block after
+	// maxBuffer searches.
+	for i := 0; i < repeatedSearches; i++ {
+		if time > headTime+maxTimeBuffer {
```
Comment (shayzluf) [inline beacon-chain/powchain/block_reader.go:114]: time> headtime+x ? 
i think it shouldnt be possible with  `if time > s.latestEth1Data.BlockTime {`

Comment (nisdas) [reply beacon-chain/powchain/block_reader.go:114]: if you look in the loop we reset the head time each time, so this is to handle those cases.

Comment (shayzluf) [reply beacon-chain/powchain/block_reader.go:114]: i see `headTime = hinfo.Time`

Comment (shayzluf) [reply beacon-chain/powchain/block_reader.go:114]: nm

File: beacon-chain/powchain/block_reader_test.go [hunk 2]
```diff
@@ -185,23 +185,90 @@ func TestBlockExists_UsesCachedBlockInfo(t *testing.T) {
...
+	ctx := context.Background()
+	hd, err := testAcc.Backend.HeaderByNumber(ctx, nil)
+	require.NoError(t, err)
+	web3Service.latestEth1Data.BlockTime = hd.Time
+	// Provide an unattainable target time
+	_, err = web3Service.findLessTargetEth1Block(ctx, hd.Number, hd.Time/2)
+	require.ErrorContains(t, errSearch.Error(), err)
+
+	// Provide an attainable target time
+	bn, err := web3Service.findLessTargetEth1Block(ctx, hd.Number, hd.Time-5)
```
Comment (shayzluf) [inline beacon-chain/powchain/block_reader_test.go:238]:
```suggestion
	bn, err := web3Service.findLessTargetEth1Block(ctx, hd.Number, hd.Time-5*params.BeaconConfig().SecondsPerETH1Block)
```

Comment (nisdas) [reply beacon-chain/powchain/block_reader_test.go:238]: It is intentionally 5 seconds not (5 *block_time)

File: beacon-chain/powchain/block_reader_test.go [hunk 3]
```diff
@@ -185,23 +185,90 @@ func TestBlockExists_UsesCachedBlockInfo(t *testing.T) {
...
+	require.NoError(t, err)
+	require.NotEqual(t, hd.Number.Uint64(), bn.Uint64(), "retrieved block is less than the head")
```
Comment (rkapka) [inline beacon-chain/powchain/block_reader_test.go:240]: The error message is incorrect. It should be `retrieved block is not less than the head`.

File: beacon-chain/powchain/block_reader_test.go [hunk 4]
```diff
@@ -185,23 +185,90 @@ func TestBlockExists_UsesCachedBlockInfo(t *testing.T) {
...
+	web3Service.eth1DataFetcher = &goodFetcher{backend: testAcc.Backend}
+
+	for i := 0; i < 200; i++ {
+		testAcc.Backend.Commit()
+	}
+	ctx := context.Background()
+	hd, err := testAcc.Backend.HeaderByNumber(ctx, nil)
+	require.NoError(t, err)
+	web3Service.latestEth1Data.BlockTime = hd.Time
+	// Provide an unattainable target time with respective to head
```
Comment (rkapka) [inline beacon-chain/powchain/block_reader_test.go:262]:
```suggestion
	// Provide an unattainable target time with respect to head
```

File: beacon-chain/powchain/block_reader_test.go [hunk 5]
```diff
@@ -185,23 +185,90 @@ func TestBlockExists_UsesCachedBlockInfo(t *testing.T) {
...
+	_, err = web3Service.findMoreTargetEth1Block(ctx, big.NewInt(0).Div(hd.Number, big.NewInt(2)), hd.Time)
+	require.ErrorContains(t, errSearch.Error(), err)
+
+	// Provide an attainable target time with respective to head
```
Comment (rkapka) [inline beacon-chain/powchain/block_reader_test.go:266]:
```suggestion
	// Provide an attainable target time with respect to head
```

**Review by rkapka — CHANGES_REQUESTED** [review]
`findLessTargetEth1Block` and `findMoreTargetEth1Block` are really hard to name given that they basically do the same thing, but with a different algorithm.. I think I would use the same comment for both functions, just changing info about when the function is used.

Out of curiosity: is there any specific reason for choosing `searchThreshold = 3` and `repeatedSearches = 6`?

File: beacon-chain/powchain/block_reader.go [hunk 6]
```diff
@@ -84,41 +94,115 @@ func (s *Service) BlockTimeByHeight(ctx context.Context, height *big.Int) (uint6
...
-	for bn := head.Number; ; bn = big.NewInt(0).Sub(bn, big.NewInt(1)) {
+	if headTime >= time {
+		return s.findLessTargetEth1Block(ctx, big.NewInt(int64(estimatedBlk)), time)
+	}
+	return s.findMoreTargetEth1Block(ctx, big.NewInt(int64(estimatedBlk)), time)
+}
+
+// Performs a search to find a target eth1 block which is less than or equal to the
+// target time. This method is used when head.time >= targetTime
+func (s *Service) findLessTargetEth1Block(ctx context.Context, head *big.Int, targetTime uint64) (*big.Int, error) {
```
Comment (shayzluf) [inline beacon-chain/powchain/block_reader.go:141]:
```suggestion
func (s *Service) findLessTargetEth1Block(ctx context.Context, startBlock *big.Int, targetTime uint64) (*big.Int, error) {
```

File: beacon-chain/powchain/block_reader.go [hunk 7]
```diff
@@ -84,41 +94,115 @@ func (s *Service) BlockTimeByHeight(ctx context.Context, height *big.Int) (uint6
...
 	defer span.End()
 
-	head, err := s.eth1DataFetcher.HeaderByNumber(ctx, nil)
-	if err != nil {
-		return nil, err
+	if time > s.latestEth1Data.BlockTime {
+		return nil, errors.New("provided time is later than the current eth1 head")
+	}
+	headNumber := big.NewInt(int64(s.latestEth1Data.BlockHeight))
+	headTime := s.latestEth1Data.BlockTime
```
Comment (shayzluf) [inline beacon-chain/powchain/block_reader.go:107]:
```suggestion
	cursor := s.latestEth1Data.BlockTime
```
maybe rename it to make it clear that its not the head block time

File: beacon-chain/powchain/block_reader.go [hunk 8]
```diff
@@ -84,41 +94,115 @@ func (s *Service) BlockTimeByHeight(ctx context.Context, height *big.Int) (uint6
...
+		if info.Time <= targetTime {
+			return info.Number, nil
+		}
+	}
+	return nil, errSearch
+}
+
+// Performs a search to find a target eth1 block which is the the block which
+// is just less than to the target time. This method is used when head.time < targetTime
+func (s *Service) findMoreTargetEth1Block(ctx context.Context, head *big.Int, targetTime uint64) (*big.Int, error) {
```
Comment (shayzluf) [inline beacon-chain/powchain/block_reader.go:162]:
```suggestion
func (s *Service) findMoreTargetEth1Block(ctx context.Context, startBlockNum *big.Int, targetTime uint64) (*big.Int, error) {
```

File: beacon-chain/powchain/block_reader.go [hunk 9]
```diff
@@ -84,41 +94,115 @@ func (s *Service) BlockTimeByHeight(ctx context.Context, height *big.Int) (uint6
...
 func (s *Service) BlockNumberByTimestamp(ctx context.Context, time uint64) (*big.Int, error) {
 	ctx, span := trace.StartSpan(ctx, "beacon-chain.web3service.BlockByTimestamp")
 	defer span.End()
 
-	head, err := s.eth1DataFetcher.HeaderByNumber(ctx, nil)
-	if err != nil {
-		return nil, err
+	if time > s.latestEth1Data.BlockTime {
+		return nil, errors.New("provided time is later than the current eth1 head")
+	}
```
[the remark this answers is not in the document]
Comment (nisdas) [reply beacon-chain/powchain/block_reader.go:105]: This doesn't return eth1data it returns a block number, if you have a block time that doesn't exist yet there is nothing to return so an error is why it returned.

File: beacon-chain/powchain/block_reader.go [hunk 10]
```diff
@@ -7,10 +7,20 @@ import (
...
 	"github.com/prysmaticlabs/prysm/shared/traceutil"
 	"go.opencensus.io/trace"
 )
 
+// searchThreshold to apply for when searching for blocks of a particular time. If the buffer
+// is exceeded we recalibrate the search again.
+const searchThreshold = 3
+
+// amount of times we repeat a failed search till is satisfies the conditional.
+const repeatedSearches = 2 * searchThreshold
```
Comment (shayzluf) [inline beacon-chain/powchain/block_reader.go:20]:
```suggestion
const repeatedSearches = 8 * searchThreshold
```
just to be sure we don't miss a block due to abnormality in block producing time

**Comment by rkapka** [comment]
`findLessTargetEth1Block` and `findMoreTargetEth1Block` are really hard to name given that they basically do the same thing, but with a different algorithm.. I think I would use the same comment for both functions, just changing info about when the function is used.

Out of curiosity: is there any specific reason for choosing `searchThreshold = 3` and `repeatedSearches = 6`?

File: beacon-chain/powchain/block_reader_test.go (hunk 2 above)
Comment (shayzluf) [reply beacon-chain/powchain/block_reader_test.go:238]: i thought it will be better to check the boundary

Comment (nisdas) [reply beacon-chain/powchain/block_reader_test.go:238]: which boundary would that be ?

File: beacon-chain/powchain/block_reader.go (hunk 10 above)
Comment (nisdas) [reply beacon-chain/powchain/block_reader.go:20]: I have let our searches be boundless, so we dont run into this.

Comment (nisdas) [reply beacon-chain/powchain/block_reader.go:20]: Context deadlines will be the fallback here, arbitary limits might allow some edge cases to slip through

**Review by shayzluf — APPROVED** [review]
maybe it will be nice to add a metric for average i and bn

**Review by rkapka — DISMISSED** [review]

**Review by farazdagi — APPROVED** [review]
Checked out and tried locally, lgtm.
Note: I've removed unused variable `errSearch`.

**Review by rkapka — DISMISSED** [review]
@farazdagi removed one line too many

**Review by farazdagi — APPROVED** [review]

**Comment by prestonvanloon** [comment]
@nisdas could this PR be causing eth1 data voting issues? We are seeing some potential issues in Medalla where Prysm proposers seem to be producing a random vote rather than a real block hash.

**Pull request merged**